### ISSUE TITLE
Modifications for training on new dataset and "Skipping loading of weights for layer due to mismatch in shape" error

### ISSUE THREAD

**Issue by JudyMRSD** [qqwweee/keras-yolo3#34]
Thank you very much for this great repo!  I have a few questions for training. 

### Question 1: 
I have a pretrained weight ('five_class_weight.h5) with 5 classes (car, bus, minibus, cyclist, truck) , and I'm trying to fine tune the result on 40 images with 2 classes (car, bus). Is there other changes to the files that I missed?

The pretrained weights could detect 80% cars my dataset. I used this dataset as training dataset,  but after fine tune for 300 epochs it cannot detect any cars in the training dataset. My guess is it won't get worse than before if I'm doing it correctly. 

First I converted weights for the 5 class model:
```
python convert.py five_class.cfg five_class.weights model_data/five_class_weight.h5
```
I'm still using the 5 classes for classes path, since this network will have a smaller change on the last layer. 
```
train.py

```javascript
def _main():
    anchors_path = 'model_data/yolo_anchors.txt'
    classes_path = 'model_data/five_classes.txt'


def get_training_data(annotation_path, data_path, input_shape, max_boxes=100, load_previous=True):
    if load_pretrained:
           weights_path = os.path.join('model_data', 'five_class_weight.h5')
```
### Question 2: 
What are the difference between my_yolo.h5 and trained_weights.h5 ? Which one is should I use when use yolo.py for detection? 
```javascript
def _main():
    ......
    output_path = 'model_data/my_yolo.h5'

def train(model, image_data, y_true, log_dir='logs/'):
   ......
    model.save_weights(log_dir + 'trained_weights.h5')
```
yolo.py
```
self.model_path = 'model_data/my_yolo.h5'
self.anchors_path = 'model_data/yolo_anchors.txt'
self.classes_path = 'model_data/five_classes.txt'
```
### Question 3: During training, I got the following warning. What does these warning mean? Would these affect the training result?
```
/home/jinz1/anaconda/envs/keras-yolo3-noaug/lib/python3.6/site-packages/keras/engine/topology.py:3462: UserWarning: Skipping loading of weights for layer conv2d_59 due to mismatch in shape ((1, 1, 1024, 75) vs (255, 1024, 1, 1)).
  weight_values[i].shape))
/home/jinz1/anaconda/envs/keras-yolo3-noaug/lib/python3.6/site-packages/keras/engine/topology.py:3462: UserWarning: Skipping loading of weights for layer conv2d_59 due to mismatch in shape ((75,) vs (255,)).
  weight_values[i].shape))
/home/jinz1/anaconda/envs/keras-yolo3-noaug/lib/python3.6/site-packages/keras/engine/topology.py:3462: UserWarning: Skipping loading of weights for layer conv2d_67 due to mismatch in shape ((1, 1, 512, 75) vs (255, 512, 1, 1)).
  weight_values[i].shape))
/home/jinz1/anaconda/envs/keras-yolo3-noaug/lib/python3.6/site-packages/keras/engine/topology.py:3462: UserWarning: Skipping loading of weights for layer conv2d_67 due to mismatch in shape ((75,) vs (255,)).
  weight_values[i].shape))
/home/jinz1/anaconda/envs/keras-yolo3-noaug/lib/python3.6/site-packages/keras/engine/topology.py:3462: UserWarning: Skipping loading of weights for layer conv2d_75 due to mismatch in shape ((1, 1, 256, 75) vs (255, 256, 1, 1)).
  weight_values[i].shape))
/home/jinz1/anaconda/envs/keras-yolo3-noaug/lib/python3.6/site-packages/keras/engine/topology.py:3462: UserWarning: Skipping loading of weights for layer conv2d_75 due to mismatch in shape ((75,) vs (255,)).
  weight_values[i].shape))
Train on 32 samples, validate on 4 samples
Epoch 1/300
32/32 [==============================] - 3s 108ms/step - loss: 4225.3423 - val_loss: 4649.4351
...
Epoch 300/300
32/32 [==============================] - 1s 23ms/step - loss: 27.5249 - val_loss: 49.5864
```
I trained for more epochs later, but the detection result has lower accuracy than the pretrained model. After fine tune, 4 out of 10 cars are detected and labeled as "truck". Before fine tune, 80% cars can be correctly detected. 
```
Epoch 1780/30000
32/32 [==============================] - 1s 25ms/step - loss: 16.4085 - val_loss: 38.6537
Epoch 01780: early stopping
```

Sorry for this long post of questions. Thank you very much for your help!

**Comment by qqwweee** [comment]
Q3, mismatch warning is OK when you retrain a new model that has different classes. But, as you say, you fine-tune the model, so output layers' weights need to be loaded rather than ignored because of mismatch. On the other hand, your model has 5 classes, so output conv layers have 3\*(5+5)=30 filters. But the warning message says 75 mismatches 255.
Q2, save_weights does not save the information about model structure so it cannot be used for load_model.
Q1, as there is confusion in Q3, the reason is hard to say. By the way, if you fine-tune 2 classed on 5 classes model, undesired weight decay will occur because of L2 regularization.

**Comment by JudyMRSD** [comment]
Thank you very much! Your answer gives great insight and now I know where went wrong, for some reason the pretrained model has 20 classes as last layer but it only classifies 5 classes of objects, so 75/3=25=20+5 is for 20 classes. I'll try agin : )

**Comment by vedhas** [comment]
@JudyMRSD 
Generic question :slightly_smiling_face:  
Q0: Were you able to solve these issues? That is, how exactly did you train/fine-tune the pretrained model to focus on only two of the total five/twenty classes? 

Specific questions:

Q1: Do I need to reassign/reorder class-labels e.g. 0. minibus, 1. cycle 2. car 3. truck 4. bus to 0. bus 1. car in `classes.txt` and also the annotation txt file? 

Q2: Can you explain "the pretrained model has 20 classes as last layer but it only classifies 5 classes of objects"? 

Q3: When fine-tuning a pretrained model for a subset of all classes, is it possible to assign 0 loss contributions to come from other classes (x,y,w,h and c values) that we are not interested in? That way, we would not need to change the topology of the network at all, and could continue fine-tuning the model to focus on one/a few classes at a time, at will. I remember trying this sometime back, but I kept getting some sizing issues/placeholder assignment issues. Kindly let me know if you tried this and succeeded.
